lib/rouge/lexers/armasm.rb

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class ArmAsm < RegexLexer
+      title "ArmAsm"
+      desc "Arm assembly syntax"
+      tag 'armasm'
+      filenames '*.s'
+
+      state :root do
+        rule /\n/, Text
+        rule /^[ \t]*#[ \t]*((define|elif|else|endif|error|if|ifdef|ifndef|include|line|pragma|undef|warning)[ \t].*)?\n/, Comment::Preproc
```

Comment:
There's a great deal of variation in the lexers in Rouge but we're trying to get new ones to be more consistent. Could you pull these keywords out and put them behind a class method that memoises the result? Something like:

```rb 
def self.keywords
  @keywords ||= Set.new %w(define elif … warning)
end
```

You then reference this as `self.class.keywords` in your rules.

Reply:
Thanks for taking a look. I've taken a brief look at how self.keywords is used in other lexers (awk.rb and c.rb because I happen to be familiar with the languages). It looks to me as though it's only useful for deciding what action to take (using cascading if/elseif tests) after a particular rule has already been matched, where a relatively general regex (called `id` in the examples I looked at) is used to decide whether to match the rule.

Is there any way to write an else clause which says "in this eventuality, please go on and test the remaining rules"? 

This first rule that you quoted is quite a good example of why this would be needed. The C preprocessor was a late addition to the syntax, probably at least 10 years after it was initially designed. Therefore, the rest of the syntax wasn't designed with it in mind, and a line such as

```
        #       foo
```

was already perfectly valid syntax (it means allocate an amount of space calculated using the assembly-time expression to the right of the hash). When the preprocessor was added, this meant there was an ambiguity in lines such as

```
        #       warning
```
I disambiguate by explicitly listing the complete set of keywords that are permitted by the C preprocessor, and that affects the way the whole line is parsed.

Reply:
Ah, yes, they are sometimes used like that but that's not what I had in mind here. Sorry, that wasn't very clear.

Basically, instead of this:

```rb
state :example
  rule /a|simple|example/, Some::Token
end
```
you write this:

```rb
self.keywords
  @keywords ||= Set.new %(a simple example)
end

state :example
  rule /#{self.class.keywords.join("|")}/, Some::Token
end
```

In your case, this doesn't actually have a performance improvement (if anything, it's a detriment) but it can be important in other lexers and bringing some conformity to overall lexer implementation is a long-term goal. It also does make the keywords easier to review and maintain, which is a plus.

Reply:
Alternatively, you can just use `Regexp.union` for generating simple regexps from arrays:
```ruby
state :example do
  rule Regexp.union(self.class.keywords), Some::Token
end
```